File: src/main/java/guru/xgm/language/jass/openapi/actionSystem/convert/Jass2AnyVisitor.java
package guru.xgm.language.jass.openapi.actionSystem.convert;

import com.intellij.psi.PsiElement;
import guru.xgm.language.jass.psi.*;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Objects;

public abstract class Jass2AnyVisitor extends JassVisitor {
    public Jass2AnyVisitor() {
        keywords = new HashSet<>();
    }

    private static final String keywordSuffix = "_FUCKING_KEYWORD";

    public String getSafeName(String name) {
        if (keywords.contains(name)) return name + keywordSuffix;
        return name;
    }

    public void appendSafeName(String name) {
        stringBuffer.append(name);
        if (keywords.contains(name)) stringBuffer.append(keywordSuffix);
    }

    public HashSet<String> keywords;

    public StringBuffer stringBuffer = new StringBuffer();

    public void appendSingleLineComment(String comment) {
        stringBuffer.append("//").append(comment).append("\n");
    }

    public void appendBlockLineComment(String comment) {
        stringBuffer.append("/*").append(comment).append("*/");
    }

    public void appendStatementLineEnd() {
        stringBuffer.append(";\n");
    }

    public void appendString(@NotNull String str) {
        stringBuffer.append(str);
    }

    public void appendReal(@NotNull String real) {
        stringBuffer.append(real);
    }

    public void appendNull() {
        stringBuffer.append("null");
    }

    public void appendFunctionAsCode(@NotNull String name) {
        stringBuffer.append(name);
    }

    public @NotNull String getConvertedTypeName(@NotNull String type) {
        return type;
    }

    // --- typedef
    public void appendTypeDef(String name, String base) {
        appendSingleLineComment(name + " extends " + base);
    }

    @Override
    public void visitTypeDef(@NotNull JassTypeDef o) {
        final var name = o.getTypeName();
        final var base = o.getTypeNameBase();
        if (name == null || base == null) return;
        appendTypeDef(name.getText(), base.getText());
    }

    // --- globals
    @Override
    public void visitGlob(@NotNull JassGlob o) {
        appendSingleLineComment("globals");
        o.acceptChildren(this);
        appendSingleLineComment("endglobals");
    }

    // --- variable
    public abstract void appendVar(boolean constant, boolean global, boolean array, @NotNull String type, String name, @Nullable JassExpr expr);

    private void appendVarPrepare(boolean constant, boolean global, @NotNull JassVar var) {
        appendVar(constant, global, var.getArray() != null, getConvertedTypeName(var.getTypeName().getText()), getSafeName(var.getId().getText()), var.getExpr());
    }

    @Override
    public void visitGvar(@NotNull JassGvar o) {
        appendVarPrepare(o.getConstant() != null, true, o.getVar());
    }

    @Override
    public void visitLvarStmt(@NotNull JassLvarStmt o) {
        appendVarPrepare(false, false, o.getVar());
    }

    // --- function
    public abstract void appendFunction(@Nullable String returns, String name, @NotNull List<JassParam> params, @NotNull List<JassStmt> statements);

    @Override
    public void visitFun(@NotNull JassFun o) {
        final JassFunRet ret = o.getFunRet();
        final @Nullable var rettype = ret != null ? ret.getTypeName() : null;

        final var name = o.getId();

        final var take = o.getFunTake();
        List<JassParam> params = new ArrayList<>();
        if (take != null) {
            final var list = take.getParamList();
            if (list != null) params = list.getParamList();
        }

        appendFunction(
                ret == null || rettype == null || ret.getNothing() != null ? null : getConvertedTypeName(rettype.getText()),
                name == null ? "" : getSafeName(name.getText()),
                params,
                o.getStmtList()
        );
    }

    // --- statement

    // stmt list
    @Override
    public void visitStmt(@NotNull JassStmt o) {
        o.acceptChildren(this);
    }

    // return
    @Override
    public void visitReturnStmt(@NotNull JassReturnStmt o) {
        stringBuffer.append("return ");
        acceptExpr(o.getExpr());
        appendStatementLineEnd();
    }

    // set
    @Override
    public void visitSetStmt(@NotNull JassSetStmt o) {
        final var id = o.getId();
        if (id != null) appendSafeName(id.getText());
        final var arr = o.getArrayAccess();
        if (arr != null) arr.accept(this);
        final var expr = o.getExpr();
        if (expr != null) {
            stringBuffer.append(" = ");
            expr.accept(this);
        }
        appendStatementLineEnd();
    }

    // call
    @Override
    public void visitCallStmt(@NotNull JassCallStmt o) {
        if (o.getDebug() != null) appendBlockLineComment("debug");
        o.getFunCall().accept(this);
        appendStatementLineEnd();
    }

    @Override
    public void visitFunCall(@NotNull JassFunCall o) {
        final var name = o.getId().getText();
        final var list = o.getArgList();
        if (Objects.equals(name, "ExecuteFunc") && list != null) {
            final var exprs = list.getExprList();
            if (exprs.size() == 1) {
                final var expr = exprs.get(0).getFirstChild();
                if (expr.getNode().getElementType() == JassTypes.STRVAL) {
                    String fn = expr.getText().replaceFirst("^\"", "").replaceFirst("\"$", "");
                    if (fn.matches("^[a-zA-Z_][a-zA-Z_0-9]*$")) {
                        stringBuffer.append(fn).append("()");
                        return;
                    }
                }
            }
        }

        appendSafeName(name);
        stringBuffer.append("(");
        if (list != null) list.accept(this);

        stringBuffer.append(")");
    }

    @Override
    public void visitArgList(@NotNull JassArgList o) {
        final var list = o.getExprList();
        for (int i = 0; i < list.size(); i++) {
            list.get(i).accept(this);
            if (i < list.size() - 1) stringBuffer.append(", ");
        }
    }

    @Override
    public void visitFuncAsCode(@NotNull JassFuncAsCode o) {
        appendFunctionAsCode(getSafeName(o.getId().getText()));
    }

    // ===============

    private boolean text(@Nullable PsiElement pe) {
        if (pe == null) return false;
        stringBuffer.append(pe.getText());
        return true;
    }

    @Override
    public void visitArrayAccess(@NotNull JassArrayAccess o) {
        stringBuffer.append(o.getId().getText()).append("[");
        final var expr = o.getExpr();
        if (expr != null) expr.accept(this);
        stringBuffer.append("]");
    }

    // --- expression
    public void acceptExpr(@Nullable JassExpr expr) {
        if (expr == null) return;
        expr.accept(this);
    }

    public void appendExprWithPrefixOp(@Nullable JassExpr expr, String op) {
        if (expr == null) return;
        stringBuffer.append(" ").append(op).append(" ");
        expr.accept(this);
    }

    public void appendExprListConcatByOperator(List<JassExpr> list, String op) {
        list.get(0).accept(this);
        stringBuffer.append(" ").append(op).append(" ");
        list.get(1).accept(this);
    }

    // primary
    @Override
    public void visitPrimExpr(@NotNull JassPrimExpr o) {
        if (text(o.getTrue())) return;
        if (text(o.getFalse())) return;
        final var tnull = o.getNull();
        if (tnull != null) {
            appendNull();
            return;
        }
        final var id = o.getId();
        if (id != null) {
            appendSafeName(id.getText());
            return;
        }
        if (text(o.getIntval())) return;
        final var str = o.getStrval();
        if (str != null) {
            appendString(str.getText());
            return;
        }
        final var raw = o.getRawval();
        if (raw != null) {
            stringBuffer.append("FourCC(\"").append(raw.getText().replace("'", "")).append("\")");
            return;
        }
        final var hex = o.getHexval();
        if (hex != null) {
            stringBuffer.append(hex.getText().replace("$", "0x"));
            return;
        }
        final var real = o.getRealval();
        if (real != null) {
            appendReal(real.getText());
            return;
        }
        o.acceptChildren(this);
    }

    @Override
    public void visitMulUnExpr(@NotNull JassMulUnExpr o) {
        appendExprWithPrefixOp(o.getExpr(), "*");
    }

    @Override
    public void visitDivUnExpr(@NotNull JassDivUnExpr o) {
        appendExprWithPrefixOp(o.getExpr(), "/");
    }

    @Override
    public void visitPlusUnExpr(@NotNull JassPlusUnExpr o) {
        appendExprWithPrefixOp(o.getExpr(), "+");
    }

    @Override
    public void visitMinusUnExpr(@NotNull JassMinusUnExpr o) {
        appendExprWithPrefixOp(o.getExpr(), "-");
    }

    @Override
    public void visitNotExpr(@NotNull JassNotExpr o) {
        appendExprWithPrefixOp(o.getExpr(), "!");
    }

    @Override
    public void visitEqExpr(@NotNull JassEqExpr o) {
        appendExprListConcatByOperator(o.getExprList(), "==");
    }

    @Override
    public void visitNeqExpr(@NotNull JassNeqExpr o) {
        appendExprListConcatByOperator(o.getExprList(), "!=");
    }

    @Override
    public void visitLtExpr(@NotNull JassLtExpr o) {
        appendExprListConcatByOperator(o.getExprList(), "<");
    }

    @Override
    public void visitLtEqExpr(@NotNull JassLtEqExpr o) {
        appendExprListConcatByOperator(o.getExprList(), "<=");
    }

    @Override
    public void visitGtExpr(@NotNull JassGtExpr o) {
        appendExprListConcatByOperator(o.getExprList(), ">");
    }

    @Override
    public void visitGtEqExpr(@NotNull JassGtEqExpr o) {
        appendExprListConcatByOperator(o.getExprList(), ">=");
    }

    @Override
    public void visitParenExpr(@NotNull JassParenExpr o) {
        stringBuffer.append("(");
        acceptExpr(o.getExpr());
        stringBuffer.append(")");
    }

    @Override
    public void visitAndExpr(@NotNull JassAndExpr o) {
        appendExprListConcatByOperator(o.getExprList(), "&&");
    }

    @Override
    public void visitOrExpr(@NotNull JassOrExpr o) {
        appendExprListConcatByOperator(o.getExprList(), "||");
    }

    @Override
    public void visitPlusExpr(@NotNull JassPlusExpr o) {
        appendExprListConcatByOperator(o.getExprList(), "+");
    }

    @Override
    public void visitMinusExpr(@NotNull JassMinusExpr o) {
        appendExprListConcatByOperator(o.getExprList(), "-");
    }

    @Override
    public void visitMulExpr(@NotNull JassMulExpr o) {
        appendExprListConcatByOperator(o.getExprList(), "*");
    }

    @Override
    public void visitDivExpr(@NotNull JassDivExpr o) {
        appendExprListConcatByOperator(o.getExprList(), "/");
    }


}

File: src/main/kotlin/raft/war/language/jass/openapi/actionSystem/convert/Jass2AnyVisitor.kt
package raft.war.language.jass.openapi.actionSystem.convert

import com.intellij.psi.PsiElement
import raft.war.language.jass.codeInspection.number.JassRawcode
import raft.war.language.jass.psi.*

abstract class Jass2AnyVisitor : raft.war.language.jass.psi.JassVisitor() {
    fun getSafeName(name: String): String {
        if (keywords.contains(name)) return name + keywordSuffix
        return name
    }

    fun appendSafeName(name: String) {
        stringBuffer.append(name)
        if (keywords.contains(name)) stringBuffer.append(keywordSuffix)
    }

    @JvmField
    var keywords: HashSet<String> = HashSet()

    @JvmField
    var stringBuffer: StringBuffer = StringBuffer()

    open fun appendSingleLineComment(comment: String?) {
        stringBuffer.append("//").append(comment).append("\n")
    }

    open fun appendBlockLineComment(comment: String?) {
        stringBuffer.append("/*").append(comment).append("*/")
    }

    open fun appendStatementLineEnd() {
        stringBuffer.append(";\n")
    }

    fun appendString(str: String) {
        stringBuffer.append(str)
    }

    open fun appendReal(real: String) {
        stringBuffer.append(real)
    }

    open fun appendNull() {
        stringBuffer.append("null")
    }

    open fun appendFunctionAsCode(name: String) {
        stringBuffer.append(name)
    }

    open fun getConvertedTypeName(type: String): String {
        return type
    }

    // --- typedef
    open fun appendTypeDef(name: String, base: String) {
        appendSingleLineComment("$name extends $base")
    }

    override fun visitTypeDef(o: raft.war.language.jass.psi.JassTypeDef) {
        val name = o.typeName
        val base = o.typeNameBase
        if (name == null || base == null) return
        appendTypeDef(name.text, base.text)
    }

    // --- globals
    override fun visitGlob(o: raft.war.language.jass.psi.JassGlob) {
        appendSingleLineComment("globals")
        o.acceptChildren(this)
        appendSingleLineComment("endglobals")
    }

    // --- variable
    abstract fun appendVar(
        constant: Boolean,
        global: Boolean,
        array: Boolean,
        type: String,
        name: String?,
        expr: raft.war.language.jass.psi.JassExpr?
    )

    private fun appendVarPrepare(constant: Boolean, global: Boolean, `var`: raft.war.language.jass.psi.JassVar) {
        appendVar(
            constant,
            global,
            `var`.array != null,
            getConvertedTypeName(`var`.typeName.text),
            getSafeName(`var`.id.text),
            `var`.expr
        )
    }

    override fun visitGvar(o: raft.war.language.jass.psi.JassGvar) {
        appendVarPrepare(o.constant != null, true, o.getVar())
    }

    override fun visitLvarStmt(o: raft.war.language.jass.psi.JassLvarStmt) {
        appendVarPrepare(false, false, o.getVar())
    }

    // --- function
    abstract fun appendFunction(returns: String?, name: String?, params: List<raft.war.language.jass.psi.JassParam?>, statements: List<raft.war.language.jass.psi.JassStmt?>)

    override fun visitFun(o: raft.war.language.jass.psi.JassFun) {
        val ret = o.funRet
        val rettype = ret?.typeName

        val name = o.id

        val take = o.funTake
        var params: List<raft.war.language.jass.psi.JassParam?> = ArrayList()
        if (take != null) {
            val list = take.paramList
            if (list != null) params = list.paramList
        }

        appendFunction(
            if (ret == null || rettype == null || ret.nothing != null) null else getConvertedTypeName(rettype.text),
            if (name == null) "" else getSafeName(name.text),
            params,
            o.stmtList
        )
    }

    // --- statement
    // stmt list
    override fun visitStmt(o: raft.war.language.jass.psi.JassStmt) {
        o.acceptChildren(this)
    }

    // return
    override fun visitReturnStmt(o: raft.war.language.jass.psi.JassReturnStmt) {
        stringBuffer.append("return ")
        acceptExpr(o.expr)
        appendStatementLineEnd()
    }

    // set
    override fun visitSetStmt(o: raft.war.language.jass.psi.JassSetStmt) {
        val id = o.id
        if (id != null) appendSafeName(id.text)
        val arr = o.arrayAccess
        arr?.accept(this)
        val expr = o.expr
        if (expr != null) {
            stringBuffer.append(" = ")
            expr.accept(this)
        }
        appendStatementLineEnd()
    }

    // call
    override fun visitCallStmt(o: raft.war.language.jass.psi.JassCallStmt) {
        if (o.debug != null) appendBlockLineComment("debug")
        o.funCall.accept(this)
        appendStatementLineEnd()
    }

    override fun visitFunCall(o: raft.war.language.jass.psi.JassFunCall) {
        val name = o.id.text
        val list = o.argList
        if (name == "ExecuteFunc" && list != null) {
            val exprs = list.exprList
            if (exprs.size == 1) {
                val expr = exprs[0].firstChild
                if (expr.node.elementType === raft.war.language.jass.psi.JassTypes.STRVAL) {
                    val fn = expr.text.replaceFirst("^\"".toRegex(), "").replaceFirst("\"$".toRegex(), "")
                    if (fn.matches("^[a-zA-Z_][a-zA-Z_0-9]*$".toRegex())) {
                        stringBuffer.append(fn).append("()")
                        return
                    }
                }
            }
        }

        appendSafeName(name)
        stringBuffer.append("(")
        list?.accept(this)
        stringBuffer.append(")")
    }

    override fun visitArgList(o: raft.war.language.jass.psi.JassArgList) {
        val list = o.exprList
        for (i in list.indices) {
            list[i].accept(this)
            if (i < list.size - 1) stringBuffer.append(", ")
        }
    }

    override fun visitFuncAsCode(o: raft.war.language.jass.psi.JassFuncAsCode) {
        appendFunctionAsCode(getSafeName(o.id.text))
    }

    // ===============
    private fun text(pe: PsiElement?): Boolean {
        if (pe == null) return false
        stringBuffer.append(pe.text)
        return true
    }

    override fun visitArrayAccess(o: raft.war.language.jass.psi.JassArrayAccess) {
        stringBuffer.append(o.id.text).append("[")
        val expr = o.expr
        expr?.accept(this)
        stringBuffer.append("]")
    }

    // --- expression
    fun acceptExpr(expr: raft.war.language.jass.psi.JassExpr?) {
        if (expr == null) return
        expr.accept(this)
    }

    fun appendExprWithPrefixOp(expr: raft.war.language.jass.psi.JassExpr?, op: String?) {
        if (expr == null) return
        stringBuffer.append(" ").append(op).append(" ")
        expr.accept(this)
    }

    fun appendExprListConcatByOperator(list: List<raft.war.language.jass.psi.JassExpr>, op: String?) {
        list[0].accept(this)
        stringBuffer.append(" ").append(op).append(" ")
        list[1].accept(this)
    }

    open fun appendRawcode(raw: JassRawcode) {
        stringBuffer.append("FourCC(\"").append(raw.strval).append("\")")
    }

    // primary
    override fun visitPrimExpr(o: raft.war.language.jass.psi.JassPrimExpr) {
        if (text(o.getTrue())) return
        if (text(o.getFalse())) return
        val tnull = o.getNull()
        if (tnull != null) {
            appendNull()
            return
        }
        val id = o.id
        if (id != null) {
            appendSafeName(id.text)
            return
        }
        if (text(o.intval)) return
        val str = o.strval
        if (str != null) {
            appendString(str.text)
            return
        }
        val raw = o.rawval
        if (raw != null) {
            appendRawcode(JassRawcode(raw))
            return
        }
        val hex = o.hexval
        if (hex != null) {
            stringBuffer.append(hex.text.replace("$", "0x"))
            return
        }
        val real = o.realval
        if (real != null) {
            appendReal(real.text)
            return
        }
        o.acceptChildren(this)
    }

    override fun visitMulUnExpr(o: raft.war.language.jass.psi.JassMulUnExpr) {
        appendExprWithPrefixOp(o.expr, "*")
    }

    override fun visitDivUnExpr(o: raft.war.language.jass.psi.JassDivUnExpr) {
        appendExprWithPrefixOp(o.expr, "/")
    }

    override fun visitPlusUnExpr(o: raft.war.language.jass.psi.JassPlusUnExpr) {
        appendExprWithPrefixOp(o.expr, "+")
    }

    override fun visitMinusUnExpr(o: raft.war.language.jass.psi.JassMinusUnExpr) {
        appendExprWithPrefixOp(o.expr, "-")
    }

    override fun visitNotExpr(o: raft.war.language.jass.psi.JassNotExpr) {
        appendExprWithPrefixOp(o.expr, "!")
    }

    override fun visitEqExpr(o: raft.war.language.jass.psi.JassEqExpr) {
        appendExprListConcatByOperator(o.exprList, "==")
    }

    override fun visitNeqExpr(o: raft.war.language.jass.psi.JassNeqExpr) {
        appendExprListConcatByOperator(o.exprList, "!=")
    }

    override fun visitLtExpr(o: raft.war.language.jass.psi.JassLtExpr) {
        appendExprListConcatByOperator(o.exprList, "<")
    }

    override fun visitLtEqExpr(o: raft.war.language.jass.psi.JassLtEqExpr) {
        appendExprListConcatByOperator(o.exprList, "<=")
    }

    override fun visitGtExpr(o: raft.war.language.jass.psi.JassGtExpr) {
        appendExprListConcatByOperator(o.exprList, ">")
    }

    override fun visitGtEqExpr(o: raft.war.language.jass.psi.JassGtEqExpr) {
        appendExprListConcatByOperator(o.exprList, ">=")
    }

    override fun visitParenExpr(o: raft.war.language.jass.psi.JassParenExpr) {
        stringBuffer.append("(")
        acceptExpr(o.expr)
        stringBuffer.append(")")
    }

    override fun visitAndExpr(o: raft.war.language.jass.psi.JassAndExpr) {
        appendExprListConcatByOperator(o.exprList, "&&")
    }

    override fun visitOrExpr(o: raft.war.language.jass.psi.JassOrExpr) {
        appendExprListConcatByOperator(o.exprList, "||")
    }

    override fun visitPlusExpr(o: raft.war.language.jass.psi.JassPlusExpr) {
        appendExprListConcatByOperator(o.exprList, "+")
    }

    override fun visitMinusExpr(o: raft.war.language.jass.psi.JassMinusExpr) {
        appendExprListConcatByOperator(o.exprList, "-")
    }

    override fun visitMulExpr(o: raft.war.language.jass.psi.JassMulExpr) {
        appendExprListConcatByOperator(o.exprList, "*")
    }

    override fun visitDivExpr(o: raft.war.language.jass.psi.JassDivExpr) {
        appendExprListConcatByOperator(o.exprList, "/")
    }


    companion object {
        private const val keywordSuffix = "_FUCKING_KEYWORD"
    }
}
